core/protocolupgrade/engine.go

```diff
@@ -221,34 +220,47 @@ func (e *Engine) BeginBlock(ctx context.Context, blockHeight uint64) {
-				e.broker.Send(events.NewProtocolUpgradeProposalEvent(ctx, pup.blockHeight, pup.vegaReleaseTag, pup.dataNodeReleaseTag, pup.approvers(), eventspb.ProtocolUpgradeProposalStatus_PROTOCOL_UPGRADE_PROPOSAL_STATUS_REJECTED))
+				e.log.Info("protocol upgrade rejected", logging.String("vega-release-tag", pup.vegaReleaseTag), logging.Uint64("upgrade-block-height", pup.blockHeight))
+				e.broker.Send(events.NewProtocolUpgradeProposalEvent(ctx, pup.blockHeight, pup.vegaReleaseTag, pup.approvers(), eventspb.ProtocolUpgradeProposalStatus_PROTOCOL_UPGRADE_PROPOSAL_STATUS_REJECTED))
 			}
 		}
 	}
+	e.lock.Lock()
+
 	if accepted != nil {
-		ID := protocolUpgradeProposalID(accepted.blockHeight, accepted.vegaReleaseTag, accepted.dataNodeReleaseTag)
-		delete(e.activeProposals, ID)
-		delete(e.events, ID)
-		e.broker.Send(events.NewProtocolUpgradeProposalEvent(ctx, accepted.blockHeight, accepted.vegaReleaseTag, accepted.dataNodeReleaseTag, accepted.approvers(), eventspb.ProtocolUpgradeProposalStatus_PROTOCOL_UPGRADE_PROPOSAL_STATUS_APPROVED))
-		e.lock.Lock()
 		e.upgradeStatus.AcceptedReleaseInfo = &types.ReleaseInfo{
 			VegaReleaseTag:     accepted.vegaReleaseTag,
-			DatanodeReleaseTag: accepted.dataNodeReleaseTag,
 			UpgradeBlockHeight: accepted.blockHeight,
 		}
-		e.lock.Unlock()
-		e.log.Info("protocol upgrade agreed", logging.String("release-tag", accepted.vegaReleaseTag), logging.String("datanode-release-tag", accepted.dataNodeReleaseTag), logging.Uint64("upgrade-block-height", accepted.blockHeight))
-
-		// if a proposal has been accepted we are auto rejecting all other proposals
-		for _, ID := range proposalIDs {
-			if pup, ok := e.activeProposals[ID]; ok {
-				e.log.Info("protocol upgrade rejected", logging.String("vega-release-tag", pup.vegaReleaseTag), logging.String("datanode-release-tag", pup.dataNodeReleaseTag), logging.Uint64("upgrade-block-height", pup.blockHeight))
-				e.broker.Send(events.NewProtocolUpgradeProposalEvent(ctx, pup.blockHeight, pup.vegaReleaseTag, pup.dataNodeReleaseTag, pup.approvers(), eventspb.ProtocolUpgradeProposalStatus_PROTOCOL_UPGRADE_PROPOSAL_STATUS_REJECTED))
-				delete(e.activeProposals, ID)
-				delete(e.events, ID)
-			}
+	} else {
+		e.upgradeStatus.AcceptedReleaseInfo = &types.ReleaseInfo{}
+	}
+	e.lock.Unlock()
+}
+
+// Cleanup is called by the abci before the final snapshot is taken to clear remaining state. It emits events for the accepted and rejected proposals.
+func (e *Engine) Cleanup(ctx context.Context) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	proposalIDs := make([]string, 0, len(e.activeProposals))
+	for k := range e.activeProposals {
+		proposalIDs = append(proposalIDs, k)
+	}
+	sort.Strings(proposalIDs)
+	for _, ID := range proposalIDs {
+		pup := e.activeProposals[ID]
+		if e.isAccepted(pup) {
+			ID := protocolUpgradeProposalID(pup.blockHeight, pup.vegaReleaseTag)
```

Comment:
I don't think we need to get the ID when we already have it. The getting and sorting of the proposalIDs could be made into a nice fuinction too given we do that somewhere else in this file:
```
for _, ID := range e.getProposalIDs() {
	pup := e.activeProposals[ID]
	status := eventspb.ProtocolUpgradeProposalStatus_PROTOCOL_UPGRADE_PROPOSAL_STATUS_APPROVED

	if !e.isAccepted(pup) {
		e.log.Info("protocol upgrade rejected", logging.String("vega-release-tag", pup.vegaReleaseTag), logging.Uint64("upgrade-block-height", pup.blockHeight))
		status = eventspb.ProtocolUpgradeProposalStatus_PROTOCOL_UPGRADE_PROPOSAL_STATUS_REJECTED
	}

	e.broker.Send(events.NewProtocolUpgradeProposalEvent(ctx, pup.blockHeight, pup.vegaReleaseTag, pup.approvers(), status))
	delete(e.activeProposals, ID)
	delete(e.events, ID)
}
```